fix: AuthenticationRefresher wrapper solution

raykrueger opened the pull request:
This fixes/relates to #2438 and #290

This commit introduces the AuthenticationRefresher as an implementation of Authentication. It's purpose is to warp another Authentication instance and refresh it very n seconds by calling `provide` on the wrapped Authentication object.

This code is intedned to be used a solution to issue #2438.

Any Authentication object that provides credentials, that must be refreshed, to the ApiClient do not have a way to do so.

There are two ways to use the AuthenticationRefresher with ClientBuilder. The first option is to use the
ClientBuilder.setAuthentication(...) method. An Authentication instance can be wrapped in an AuthenticationRefresher and passed into the setAuthentication method. For example,

```
ClientBuilder.standard().setAuthentication(
  //wrap an auth and refresh it every 15 minutes (900s)
  new Authentication(someDelegateAuthenticationInstance, 900)
);
```

This is integrated up into the ClientBuilder.build() method by adding a new authenticationRefreshSeconds field and getter/setter pair. If a refresh interval is set, the Authentication object used by ClientBuilder will be wrapped with an AuthenticationRefresher.

brendandburns commented on util/src/main/java/io/kubernetes/client/util/credentials/AuthenticationRefresher.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    @Override
+    public void provide(ApiClient client) {
+        ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
```
This should be moved out into the constructor (or a static singleton)

brendandburns commented on util/src/test/java/io/kubernetes/client/util/credentials/AuthenticationRefresherTest.java:
```diff
@@ -0,0 +1,52 @@
+
+        final ApiClient client = ClientBuilder.standard().setAuthentication(wrapper).build();
+
+        Thread.sleep(3000);
```
Tests with timing are inherently flaky. Please use synchronization primitives.

dims commented:
@raykrueger is there a way/need to be able to cancel the task running in the executor?

raykrueger commented:
> @raykrueger is there a way/need to be able to cancel the task running in the executor?

Yep! I'm combining Brendan's concern with some WeakReference foo to shut the timerdown when the APIClient gets GC'd. Commit incoming!

raykrueger commented:
I started the code review thing on accident and don't know what to do about that heh.

raykrueger commented:
(Sorry for spamming, wanted a new line in the previous message). Please have a look at the latest. I've added some safety around time timer and moved it to a field. The timer is now cancelled and nulled out if the APICilent is garbage collected. Additionally I removed the flaky Thread.sleep timing from the tests in favor of Countdown Latches.

brendandburns commented on util/src/test/java/io/kubernetes/client/util/credentials/AuthenticationRefresherTest.java:
```diff
@@ -0,0 +1,98 @@
+
+        ClientBuilder.standard().setAuthentication(wrapper).build();
+
+        boolean result = latch.await(4, TimeUnit.SECONDS);
```
Don't have a timeout here. The overall test already has a timeout, so it's better to have the test timeout, vs having the await take 4 seconds and cause the test to flake.

I realize it seems unlikely that this could take 4 seconds, but on overloaded test infrastructure machines, weird scheduling things happen.

brendandburns commented on util/src/test/java/io/kubernetes/client/util/credentials/AuthenticationRefresherTest.java:
```diff
@@ -0,0 +1,98 @@
+            System.gc();
+        }
+
+        boolean result = latch.await(4, TimeUnit.SECONDS);
```
Same comment wrt. timeouts.

brendandburns commented on util/src/main/java/io/kubernetes/client/util/credentials/AuthenticationRefresher.java:
```diff
@@ -0,0 +1,96 @@
+    @Override
+    public void provide(ApiClient client) {
+
+        synchronized (this) {
```
Why do this lazily on the first call to provide vs. doing it in the contructor? Do you expect that the ApiClient will get garbage collected repeatedly but this authentication will stick around?

brendandburns replied:
Still wondering about this. Why not just create the executor once in the constructor and be done with it? This current logic seems convulted and I'm not sure I see the value of the lazy evaluation.

brendandburns commented:
Few more comments. Thanks.

raykrueger commented:
Ok, I've simplified things a bit. Let me know if this makes better sense. Thanks!

brendandburns commented on util/src/main/java/io/kubernetes/client/util/credentials/AuthenticationRefresher.java:
```diff
@@ -0,0 +1,96 @@
+        log.debug("Invoking authentication");
+        this.delegateAuthentication.provide(client);
+        // Schedule the next run.
+        scheduleRefresh(client);
```
Isn't this going to schedule a refresh every time the credential is provided? I think you want to only schedule it once.

brendandburns commented on util/src/test/java/io/kubernetes/client/util/credentials/AuthenticationRefresherTest.java:
```diff
@@ -0,0 +1,87 @@
+        assertFalse(wrapper.isRunning(), "After our unused ApiClient returned from .build() is garbage collected, the refresher should stop.");
+        verify(mockProvider, times(1)).provide(any(ApiClient.class));
+    }
+
```
If my comment above is correct and there is a bug, add a test to validate the bug and then ensure it stays fixed.

raykrueger commented:
I pushed some code a few days ago, but didn’t reach out here. Back to you sir!

yue9944882 commented:
/ok-to-test

brendandburns commented:
Sorry, I feel like I'm dragging this code review on and on, but I think this code could be simpler. Added a comment.

bryantbiggs commented:
any update on this - or is [this](https://github.com/kubernetes-client/java/pull/3406#discussion_r1616436450) comment still waiting to be resolved?

brendandburns commented:
@bryantbiggs still waiting afaik.

brendandburns commented:
btw, I dug into this deeper, this won't work correctly because when we construct the `ApiClient` the token is set once and it is never revisited.

Making token refresh work will require changes in the upstream code generator, or a patch to ApiClient.

Closing this for now since I think we will need to start with a fresh solution.

brendandburns commented:
See:
https://github.com/OpenAPITools/openapi-generator/pull/6036
